plugins/in_kubernetes_events/kubernetes_events.c

```diff
@@ -722,6 +857,43 @@ static int k8s_events_collect(struct flb_input_instance *ins,
         ctx->last_resource_version = max_resource_version;
     }
 
+    /* Now that we've done a full list, we can use the resource version and do a watch
+     * to stream updates efficiently
+     */
+    c = make_event_watch_api_request(ctx, u_conn, max_resource_version);
+    if (!c) {
+        flb_plg_error(ins, "unable to create http client");
+        goto exit;
+    }
+    initialize_http_client(c, ctx);
+
+    /* Watch will stream chunked json data, so we only send
+     * the http request, then use flb_http_get_response_data
+     * to attempt processing on available streamed data
+     */
+    b_sent = 0;
+    ret = flb_http_do_request(c, &b_sent);
+    if (ret != 0) {
+        flb_plg_error(ins, "http do request error");
+        goto exit;
+    }
+
+    ret = FLB_HTTP_MORE;
+    bytes_consumed = 0;
+    chunk_proc_ret = 0;
+    while ((ret == FLB_HTTP_MORE || ret == FLB_HTTP_CHUNK_AVAILABLE) && chunk_proc_ret == 0) {
```

Comment:
looping indefinitely inside the collect callback may not necessarily the correct way to do this. 

So far during my testing I have not seen any problems but this could result in problems with either starving out other coroutines or impeding safe shutdown.

From what I can tell there is a memory leak caused by the loop never actually exiting and therefore never calling `flb_http_client_destroy`.

```c
    ret = FLB_HTTP_MORE;
    bytes_consumed = 0;
    chunk_proc_ret = 0;
    while ((ret == FLB_HTTP_MORE || ret == FLB_HTTP_CHUNK_AVAILABLE) && chunk_proc_ret == 0) {
        ret = flb_http_get_response_data(c, bytes_consumed);
        bytes_consumed = 0;
        if( c->resp.status == 200 && ret == FLB_HTTP_CHUNK_AVAILABLE ) {
            chunk_proc_ret = process_http_chunk(ctx, c, &bytes_consumed);
        }
    }
    /* NOTE: skipping any processing after streaming socket closes */

    if (c->resp.status != 200) {
        flb_plg_warn(ins, "events watch failure, http_status=%d payload=%s", c->resp.status, c->resp.payload);
    }


    flb_plg_error(ins, "leaving collect callback"); // <- this line was added
                                                    // and never gets called.
exit:
    pthread_mutex_unlock(&ctx->lock);
    if (c) {
        flb_http_client_destroy(c);
    }
    if (u_conn) {
        flb_upstream_conn_release(u_conn);
    }
    FLB_INPUT_RETURN(0);
}
```

Here is the valgrind log:

```
==171983== 6,595 (32 direct, 6,563 indirect) bytes in 1 blocks are definitely lost in loss record 13 of 15
==171983==    at 0x484ABF3: calloc (vg_replace_malloc.c:1675)
==171983==    by 0x1F31F3: flb_calloc (flb_mem.h:95)
==171983==    by 0x1F3256: flb_kv_item_create_len (flb_kv.c:36)
==171983==    by 0x32F6D8: flb_http_add_header (flb_http_client.c:918)
==171983==    by 0x32ECE5: add_host_and_content_length (flb_http_client.c:617)
==171983==    by 0x32F1F4: flb_http_client (flb_http_client.c:763)
==171983==    by 0x446DE0: make_event_watch_api_request (kubernetes_events.c:606)
==171983==    by 0x4481F1: k8s_events_collect (kubernetes_events.c:863)
==171983==    by 0x22524B: input_pre_cb_collect (flb_input.h:517)
==171983==    by 0xFA30C6: co_init (amd64.c:117)
==171983== 
==171983== 25,184 (64 direct, 25,120 indirect) bytes in 1 blocks are definitely lost in loss record 15 of 15
==171983==    at 0x484ABF3: calloc (vg_replace_malloc.c:1675)
==171983==    by 0x2217BD: flb_calloc (flb_mem.h:95)
==171983==    by 0x225BF4: flb_input_coro_create (flb_input.h:437)
==171983==    by 0x225BF4: flb_input_coro_collect (flb_input.h:540)
==171983==    by 0x225BF4: input_collector_fd (flb_input_thread.c:158)
==171983==    by 0x22693E: engine_handle_event (flb_input_thread.c:180)
==171983==    by 0x22693E: input_thread (flb_input_thread.c:381)
==171983==    by 0x2868F1: step_callback (flb_worker.c:43)
==171983==    by 0x5348559: start_thread (pthread_create.c:447)
==171983==    by 0x53C5893: clone (clone.S:100)
==171983==
```

I see two ways this could be solved:

1. saving the flb_http_client in the context struct and releasing it during the exit callback (which might be unsafe).
2. implement the collection via flb_input_set_collector_event which would require quite a bit of work.

The in_mqtt plugin implements flb_input_set_collector_event in case you wanted something to reference.

Reply:
@pwhelan thank you for the feedback!

1) This input is threaded, I assumed this would be enough to isolate it from starving other coros, but I wasn't 100% certain. Is this enough?

2) I will look into the memory leak on shutdown, but I'd prefer to keep this as the collect time type for now as it would take a decent effort to move this to the event type.

Reply:
> This input is threaded, I assumed this would be enough to isolate it from starving other coros, but I wasn't 100% certain. Is this enough?

Running in a thread should solve that. Since it also manages to use the network API that uses coroutines I think this will work just fine.